nil pointer dereference with wallet hex restore

panic: runtime error: invalid memory address or nil pointer dereference
[signal SIGSEGV: segmentation violation code=0x1 addr=0x60 pc=0x124d53a]

goroutine 106517 [running]:
github.com/planetdecred/godcr/ui/page/components.(*WalletSelector).UnlockWalletForSyncing(0xc0038c2000, 0xc005ac21e0)
        /home/user/code/godcr/ui/page/components/wallet_selector.go:474 +0x1fa
github.com/planetdecred/godcr/ui/page/components.(*WalletSelector).startSyncing(0xc0038c2000)
        /home/user/code/godcr/ui/page/components/wallet_selector.go:439 +0x89
github.com/planetdecred/godcr/ui/page/components.(*WalletSelector).Expose(0xc0038c2000, {0x1ddf078?, 0xc003a68b00?})
        /home/user/code/godcr/ui/page/components/wallet_selector.go:70 +0xd7
github.com/planetdecred/godcr/ui/page.(*WalletDexServerSelector).OnNavigatedTo(0xc005e54180)
        /home/user/code/godcr/ui/page/wallet_dex_selector_page.go:72 +0x6d
github.com/planetdecred/godcr/app.(*PageStack).Reset(0xc001f766c0, {0xc005dd0240, 0x1, 0x1})
        /home/user/code/godcr/app/pagestack.go:153 +0x112
github.com/planetdecred/godcr/app.(*SimpleWindowNavigator).ClearStackAndDisplay(0xc001f766f0, {0x1de32e8, 0xc005e54180})
        /home/user/code/godcr/app/window.go:84 +0x8f
github.com/planetdecred/godcr/ui/page.(*CreateWallet).handlerWalletDexServerSelectorCallBacks(0xc00215b100)
        /home/user/code/godcr/ui/page/wallet_setup_page.go:492 +0x9f
github.com/planetdecred/godcr/ui/page.(*CreateWallet).HandleUserInteractions.func3()
        /home/user/code/godcr/ui/page/wallet_setup_page.go:451 +0x1d
github.com/planetdecred/godcr/ui/page/info.(*Restore).showHexRestoreModal.func1.1.2.1()
        /home/user/code/godcr/ui/page/info/restore_page.go:256 +0x12d
created by github.com/planetdecred/godcr/ui/page/info.(*Restore).showHexRestoreModal.func1.1.2
        /home/user/code/godcr/ui/page/info/restore_page.go:243 +0x125

panic: runtime error: invalid memory address or nil pointer dereference
[signal SIGSEGV: segmentation violation code=0x1 addr=0x60 pc=0x4dc7e88]

goroutine 6116 [running]:
github.com/planetdecred/godcr/ui/page/components.(*WalletSelector).UnlockWalletForSyncing(0xc005cf4000, 0xc00bea83c0)
	/Users/sirmorrison/Desktop/go projects/godcr-gio/ui/page/components/wallet_selector.go:474 +0x1c8
github.com/planetdecred/godcr/ui/page/components.(*WalletSelector).startSyncing(0xc005cf4000)
	/Users/sirmorrison/Desktop/go projects/godcr-gio/ui/page/components/wallet_selector.go:439 +0x89
github.com/planetdecred/godcr/ui/page/components.(*WalletSelector).Expose(0xc005cf4000, {0x58db558, 0xc0037e3e80})
	/Users/sirmorrison/Desktop/go projects/godcr-gio/ui/page/components/wallet_selector.go:70 +0xd7
github.com/planetdecred/godcr/ui/page.(*WalletDexServerSelector).OnNavigatedTo(0xc0058a7020)
	/Users/sirmorrison/Desktop/go projects/godcr-gio/ui/page/wallet_dex_selector_page.go:72 +0x6d
github.com/planetdecred/godcr/app.(*PageStack).Reset(0xc00002c3f0, {0xc021844200, 0x1, 0x1})
	/Users/sirmorrison/Desktop/go projects/godcr-gio/app/pagestack.go:153 +0x112
github.com/planetdecred/godcr/app.(*SimpleWindowNavigator).ClearStackAndDisplay(0xc00002c720, {0x58e4b30, 0xc0058a7020})
	/Users/sirmorrison/Desktop/go projects/godcr-gio/app/window.go:84 +0x8f
github.com/planetdecred/godcr/ui/page.(*CreateWallet).handlerWalletDexServerSelectorCallBacks(0xc0020ef100)
	/Users/sirmorrison/Desktop/go projects/godcr-gio/ui/page/wallet_setup_page.go:492 +0x9f
github.com/planetdecred/godcr/ui/page.(*CreateWallet).HandleUserInteractions.func3()
	/Users/sirmorrison/Desktop/go projects/godcr-gio/ui/page/wallet_setup_page.go:451 +0x1d
github.com/planetdecred/godcr/ui/page/info.(*Restore).showHexRestoreModal.func1.1.2.1()
	/Users/sirmorrison/Desktop/go projects/godcr-gio/ui/page/info/restore_page.go:256 +0x118
created by github.com/planetdecred/godcr/ui/page/info.(*Restore).showHexRestoreModal.func1.1.2
	/Users/sirmorrison/Desktop/go projects/godcr-gio/ui/page/info/restore_page.go:243 +0x128

reproduced @crux25

--- a/ui/page/components/wallet_selector.go
+++ b/ui/page/components/wallet_selector.go
@@ -65,10 +65,6 @@ func (ws *WalletSelector) Expose(ctx context.Context) {
 	ws.ctx = ctx
 	ws.listenForTxNotifications()
 	ws.loadWallets()
-
-	if ws.WL.MultiWallet.ReadBoolConfigValueForKey(load.AutoSyncConfigKey, false) {
-		ws.startSyncing()
-	}
 }
 
 func (ws *WalletSelector) loadWallets() {
@@ -361,7 +357,7 @@ func (ws *WalletSelector) listenForTxNotifications() {
 	ws.TxAndBlockNotificationListener = listeners.NewTxAndBlockNotificationListener()
 	err := ws.WL.MultiWallet.AddTxAndBlockNotificationListener(ws.TxAndBlockNotificationListener, true, WalletSelectorID)
 	if err != nil {
-		// log.Errorf("Error adding tx and block notification listener: %v", err)
+		log.Errorf("Error adding tx and block notification listener: %v", err)
 		return
 	}
 
@@ -375,8 +371,8 @@ func (ws *WalletSelector) listenForTxNotifications() {
 					// only if sync is completed.
 					if ws.WL.MultiWallet.IsSynced() {
 						ws.updateAccountBalance()
-						ws.ParentWindow().Reload()
 					}
+					ws.ParentWindow().Reload()
 				case listeners.NewTransaction:
 					// refresh wallets when new transaction is received
 					ws.updateAccountBalance()
@@ -431,45 +427,4 @@ func (ws *WalletSelector) updateAccountBalance() {
 			item.TotalBalance = dcrutil.Amount(totalBalance).String()
 		}
 	}
-}
-
-func (ws *WalletSelector) startSyncing() {
-	for _, wal := range ws.WL.SortedWalletList() {
-		if !wal.HasDiscoveredAccounts && wal.IsLocked() {
-			ws.UnlockWalletForSyncing(wal)
-			return
-		}
-	}
-
-	err := ws.WL.MultiWallet.SpvSync()
-	if err != nil {
-		// show error dialog
-		log.Info("Error starting sync:", err)
-	}
-}
-
-func (ws *WalletSelector) UnlockWalletForSyncing(wal *dcrlibwallet.Wallet) {
-	spendingPasswordModal := modal.NewPasswordModal(ws.Load).
-		Title(values.String(values.StrResumeAccountDiscoveryTitle)).
-		Hint(values.String(values.StrSpendingPassword)).
-		NegativeButton(values.String(values.StrCancel), func() {}).
-		PositiveButton(values.String(values.StrUnlock), func(password string, pm *modal.PasswordModal) bool {
-			go func() {
-				err := ws.WL.MultiWallet.UnlockWallet(wal.ID, []byte(password))
-				if err != nil {
-					errText := err.Error()
-					if err.Error() == dcrlibwallet.ErrInvalidPassphrase {
-						errText = values.String(values.StrInvalidPassphrase)
-					}
-					pm.SetError(errText)
-					pm.SetLoading(false)
-					return
-				}
-				pm.Dismiss()
-				ws.startSyncing()
-			}()
-
-			return false
-		})
-	ws.ParentWindow().ShowModal(spendingPasswordModal)
 }

--- a/ui/page/wallet_dex_selector_page.go
+++ b/ui/page/wallet_dex_selector_page.go
@@ -6,9 +6,11 @@ import (
 	"gioui.org/layout"
 	"gioui.org/widget"
 
+	"github.com/planetdecred/dcrlibwallet"
 	"github.com/planetdecred/godcr/app"
 	"github.com/planetdecred/godcr/ui/decredmaterial"
 	"github.com/planetdecred/godcr/ui/load"
+	"github.com/planetdecred/godcr/ui/modal"
 	"github.com/planetdecred/godcr/ui/page/components"
 	"github.com/planetdecred/godcr/ui/page/dexclient"
 	"github.com/planetdecred/godcr/ui/values"
@@ -71,6 +73,10 @@ func (pg *WalletDexServerSelector) OnNavigatedTo() {
 	pg.ctx, pg.ctxCancel = context.WithCancel(context.TODO())
 	pg.walletSelector.Expose(pg.ctx)
 	pg.dexServerSelector.Expose()
+
+	if pg.WL.MultiWallet.ReadBoolConfigValueForKey(load.AutoSyncConfigKey, false) {
+		pg.startSyncing()
+	}
 }
 
 // HandleUserInteractions is called just before Layout() to determine
@@ -222,4 +228,45 @@ func (pg *WalletDexServerSelector) layoutAddMoreRowSection(clk *decredmaterial.C
 			)
 		})
 	}
+}
+
+func (pg *WalletDexServerSelector) startSyncing() {
+	for _, wal := range pg.WL.SortedWalletList() {
+		if !wal.HasDiscoveredAccounts && wal.IsLocked() {
+			pg.UnlockWalletForSyncing(wal)
+			return
+		}
+	}
+
+	err := pg.WL.MultiWallet.SpvSync()
+	if err != nil {
+		// show error dialog
+		log.Info("Error starting sync:", err)
+	}
+}
+
+func (pg *WalletDexServerSelector) UnlockWalletForSyncing(wal *dcrlibwallet.Wallet) {
+	spendingPasswordModal := modal.NewPasswordModal(pg.Load).
+		Title(values.String(values.StrResumeAccountDiscoveryTitle)).
+		Hint(values.String(values.StrSpendingPassword)).
+		NegativeButton(values.String(values.StrCancel), func() {}).
+		PositiveButton(values.String(values.StrUnlock), func(password string, pm *modal.PasswordModal) bool {
+			go func() {
+				err := pg.WL.MultiWallet.UnlockWallet(wal.ID, []byte(password))
+				if err != nil {
+					errText := err.Error()
+					if err.Error() == dcrlibwallet.ErrInvalidPassphrase {
+						errText = values.String(values.StrInvalidPassphrase)
+					}
+					pm.SetError(errText)
+					pm.SetLoading(false)
+					return
+				}
+				pm.Dismiss()
+				pg.startSyncing()
+			}()
+
+			return false
+		})
+	pg.ParentWindow().ShowModal(spendingPasswordModal)
 }

